### PR TITLE
backupccl: ensure restore on success is run once

### PR DESCRIPTION
**Pull request by pbardea**
It seems that jobs today do not ensure that the OnSuccess callback is
called exactly once. This PR moves the cleanup stages of RESTORE,
formerly located in the OnSuccess callback to be the final steps of
Resume. This should help ensure that these stages are run once and only
once.

Release note (bug fix): Ensure that RESTORE cleanup is run exactly once.

### CODE REVIEW

File: pkg/ccl/backupccl/restore.go [hunk 1]
```diff
@@ -1579,6 +1579,9 @@ type restoreResumer struct {
 	tables      []*sqlbase.TableDescriptor
 	latestStats []*stats.TableStatisticProto
 	execCfg     *sql.ExecutorConfig
+	// A collection of stages that should only be run once.
+	statsInserted   bool
```
Comment (dt) [inline pkg/ccl/backupccl/restore.go]: don't you need to persist these, in the job detail or progress?

Comment (pbardea) [reply pkg/ccl/backupccl/restore.go]: yep. looks like this was the wrong branch.

Comment (dt) [reply pkg/ccl/backupccl/restore.go]: 👍 
I think you might need to actually update the job too to save your change to the in-mem version you're updating here. You could do it in the same txn ideally I think?

**Review by spaskob — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43933)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea and @spaskob)

___
*[pkg/ccl/backupccl/restore.go, line 1787 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43933#-LyVs_cF5rMuEhJi_KK6:-LyVs_cF5rMuEhJi_KK7:b-xk1c6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dc99cf067a9ca60535d5a4b2e1b3d8a3e76c73b/pkg/ccl/backupccl/restore.go#L1787)):*
> ```Go
> 		return nil
> 	}
> 	defer func() { details.statsInserted = true }()
> ```

I am confused why this is deferred. Should you call it only in case InsertNewStats succeeds? And also where do you actually persist the job details? Don't you need a call to job.update(...)

**Comment by spaskob** [comment]
here's an example of how to use update
https://github.com/cockroachdb/cockroach/blob/master/pkg/jobs/jobs.go#L181

File: pkg/ccl/backupccl/restore.go (hunk 1 above)
Comment (dt) [reply pkg/ccl/backupccl/restore.go]: Here's how IMPORT does something similar (ensuring that all the descs get their state changed): https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/importccl/import_stmt.go#L900

**Comment by pbardea** [comment]
RFAL

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43933)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea)

___
*[pkg/ccl/backupccl/restore.go, line 1786 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43933#-LyZviO76o3MfDt8wlko:-LyZviO8BQhL3yDhZzek:b-trzyr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f928efe0da1a6dbd9bc9fcfb3612e67cee56bd7/pkg/ccl/backupccl/restore.go#L1786)):*
> ```Go
> 	}
> 
> 	txn := r.execCfg.DB.NewTxn(ctx, "insert-stats")
> ```

Is there a reason to use this over the `.Txn(..., fn)` that takes the retry-able closure and does the error handling?

**Review by pbardea — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43933)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)

___
*[pkg/ccl/backupccl/restore.go, line 1787 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43933#-LyVs_cF5rMuEhJi_KK6:-Ly_bVJm9_K54Iio68Rf:bffvhpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dc99cf067a9ca60535d5a4b2e1b3d8a3e76c73b/pkg/ccl/backupccl/restore.go#L1787)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

I am confused why this is deferred. Should you call it only in case InsertNewStats succeeds? And also where do you actually persist the job details? Don't you need a call to job.update(...)
</blockquote></details>

Yep, as per our offline discussion I wasn't sure if returning an error from resume would ensure that it would not be run again (and therefore continuously retry to re-insert stats). But it should be safe to just return an error and mark this stage as complete in the same txn that performs the insertion. Same applies for publishing tables below.
___
*[pkg/ccl/backupccl/restore.go, line 1786 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43933#-LyZviO76o3MfDt8wlko:-Ly_bnOr9lNR85w2Rqx8:b-k8n1h8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f928efe0da1a6dbd9bc9fcfb3612e67cee56bd7/pkg/ccl/backupccl/restore.go#L1786)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Is there a reason to use this over the `.Txn(..., fn)` that takes the retry-able closure and does the error handling?
</blockquote></details>

Nope - updated to use the retry-able closure.

**Review by dt — APPROVED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43933)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)

**Review by spaskob — APPROVED** [review]
LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43933#-:-LydiBOf-JvBEshXFnm3:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)

**Comment by pbardea** [comment]
TFTRs!
bors r+

**Comment by pbardea** [comment]
`TestTruncateWhileColumnBackfill` failed and TeamCity says it looks flaky. Trying again.
bors r+

**Comment by spaskob** [comment]
yes I have seen this test flake

**Comment by pbardea** [comment]
Third time's the charm?
bors r+

**Pull request merged**